src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -0,0 +1,282 @@
+                    properties[propertyNameBytes] = (Writer<T>)property.Accept(this, state)!;
+                }
+            }
+
+            return new Writer<T?>((ref MessagePackWriter writer, T? value) =>
+            {
+                if (value is null)
+                {
+                    writer.WriteNil();
+                    return;
+                }
+
+                writer.WriteMapHeader(properties.Count);
+
+                foreach (var property in properties)
+                {
+                    writer.WriteRaw(property.Key.Span);
+                    property.Value(ref writer, value);
+                }
+            });
+        }
+
+        public override object? VisitProperty<TDeclaringType, TPropertyType>(IPropertyShape<TDeclaringType, TPropertyType> propertyShape, object? state = null)
+        {
+            var getter = propertyShape.GetGetter();
+
+            switch (propertyShape.PropertyType.Kind)
+            {
+                case TypeShapeKind.Object:
+                    var formatter = GetWriter(propertyShape.PropertyType);
+                    return new Writer<TDeclaringType>((ref MessagePackWriter writer, TDeclaringType container) => formatter(ref writer, getter(ref container)));
+                case TypeShapeKind.Enum:
+                    throw new NotImplementedException();
+                case TypeShapeKind.Nullable:
+                    throw new NotImplementedException();
+                case TypeShapeKind.Enumerable:
+                    throw new NotImplementedException();
+                case TypeShapeKind.Dictionary:
+                    throw new NotImplementedException();
+                default:
+                    throw new NotSupportedException();
+            }
+        }
+    }
+
+    private sealed class DeserializeVisitor : TypeShapeVisitor
+    {
+        internal static readonly DeserializeVisitor Instance = new();
+
+        private static readonly TypeDictionary formatters = new()
+        {
+            { typeof(string), new Reader<string>((ref MessagePackReader reader) => reader.ReadString()) },
+            { typeof(int), new Reader<int>((ref MessagePackReader reader) => reader.ReadInt32()) },
+        };
+
+        internal static Reader<T?> GetReader<T>(ITypeShape<T> typeShape)
+        {
+            return formatters.GetOrAdd<Reader<T?>>(typeShape, Instance, box => (ref MessagePackReader reader) => box.Result(ref reader));
+        }
+
+        private DeserializeVisitor() { }
+
+        private delegate void PropertySetter<T>(ref MessagePackReader reader, ref T container);
+
+        public override object? VisitObject<T>(IObjectTypeShape<T> objectShape, object? state = null)
+        {
+            SpanDictionary<byte, PropertySetter<T>> properties = objectShape.GetProperties()
+                .Where(p => p.HasSetter)
+                .ToSpanDictionary(
+                    p => Encoding.UTF8.GetBytes(p.Name), 
```

Comment:
Does message pack support any kind of escaping or case insensitivity in its property names that would make ordinal UTF-8 comparison unsuitable?